rpc,upgrade: add `NodeDialerNoBreaker` interface

cthumuluru-crdb opened the pull request:
Currently, `NodeDialer` interface only has `Dial()` method which most code paths use. But there are also code paths that use `DialNoBreaker()` method. To support those cases, this commit adds a new interface `NodeDialerNoBreaker`.

Epic: CRDB-48923
Informs: https://github.com/cockroachdb/cockroach/issues/147757
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147997)
<!-- Reviewable:end -->

cthumuluru-crdb committed:
rpc,upgrade: add `NodeDialerNoBreaker` interface

Currently, `NodeDialer` interface only has `Dial()` method which most
code paths use. But there are also code paths that use `DialNoBreaker()`
method. To support those cases, this commit adds a new interface
`NodeDialerNoBreaker`.

Epic: CRDB-48923
Informs: https://github.com/cockroachdb/cockroach/issues/147757
Release note: none

cthumuluru-crdb commented:
bors r+
